Collision detection + SFML display

alexdewar opened the pull request:
Hey,

This is just something I've been working on for the robot experiments we were going to do in the drone zone.

I've added a simple 2D collision detector. You input the dimensions of the robot and of the objects you're avoiding + how close the robot is allowed to get, then you can check if a collision has occurred or not.

I've also included a little generic SFML display for viewing (you might need to install the libs on your machine).

neworderofjamie commented:
I'm trying to do a little bit of GeNN coding this afternoon so reviewing may have to wait until next week. However, my immediate thought is do we really want bits of our code using SFML, GLFW and SDL? 

- All three let you create windows with OpenGL contexts in a cross-platform way. 
- GLFW does nothing else and is pleasantly modern (which, as that was what I wanted, is why I chose it originally)
- SDL does some other stuff and is decidedly elderly, written and C, but would let us target important platforms like Amiga...
-  SFML does a whole load of stuff and looks perfectly nice.

I think the ability to do stuff like this P.R. is probably a good thing so maybe we should port everything across to SFML? Let me know your thoughts and, if you agree, we can split that thankless task between us.

alexdewar commented:
I'm all for using SFML -- it's got a really nice interface. The only problem is that you can only do 2D stuff with it, so we couldn't use it for e.g. libantworld :disappointed: 

Norbert's ``CarDisplay`` class could be replaced with the ``SFMLRenderer`` though.

neworderofjamie commented:
According to https://www.sfml-dev.org/tutorials/2.5/window-opengl.php that would seem not to be the case!

alexdewar commented:
Ahh sorry, I see what you meant now. Yes, I guess we could drop the GLFW stuff and mix and match SFML with OpenGL.

That's not a bad idea. Replacing the GLFW bits with SFML would probably make the code more readable apart from anything.

I'm happy to divide this task up!

neworderofjamie commented:
Hmm I think that could be a little optimistic - all the OpenGL state
fiddling will remain cos that's the nature of the beast beneath but I think
sticking with one library around the outside is a good thing

On Fri, 1 Mar 2019 at 15:44, Alex Dewar <notifications@github.com> wrote:

> Ahh sorry, I see what you meant now. Yes, I guess we could drop the GLFW
> stuff and mix and match SFML with OpenGL.
>
> That's not a bad idea. Replacing the GLFW bits with SFML would probably
> make the code more readable apart from anything.
>
> I'm happy to divide this task up!
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/BrainsOnBoard/bob_robotics/pull/97#issuecomment-468708699>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGeoGqQKIzDyZKXQ_ETEZ1ny4BTkgMjuks5vSUr1gaJpZM4bZOpD>
> .
>

alexdewar commented:
Hey,

Would you mind having a quick look at this when you get a chance? It's not urgent, but I've got a couple more branches based on this (with pretty minor stuff in) that I'd also like to merge soonish.

alexdewar commented:
PS -- Not sure if you have SFML installed or not? Looks like it can't find the package.

alexdewar commented:
Actually, I've just realised it makes sense to merge some later stuff into this branch. Hang on a mo

alexdewar commented:
Ok, done. I also added the functionality we have with the SDL code into ``SFMLDisplay``. Originally I also deleted all the SDL stuff on this branch, but I've reverted that as it looks like @tenxlenx is still using it.

tenxlenx commented:
I use SDL because for some mysterious reason SFML does not work on the office computer, otherwise I prefer SFML because it has so much more functionality

neworderofjamie commented:
@alexdewar and I are good at breaking office computers - let us know when you're next in and we can fight it.

neworderofjamie commented on common/sfml_display.h:
```diff
@@ -0,0 +1,321 @@
+using namespace units::literals;
+
+template<typename LengthUnit = units::length::meter_t>
+class SFMLDisplay
```
Maybe ``SFMLAgentArena`` or ``SFMLWorld`` would be better names - display suggests to me at least that it's a lower-level thing

alexdewar replied:
:+1:

alexdewar commented on common/pose.h:
```diff
@@ -25,6 +29,8 @@ template<typename Derived>
 class PoseBase
```
Actually I am using a CRTP here... or is that not what you meant?

neworderofjamie replied:
Ahh ok, in that case ``Derived`` is probably an alright name

alexdewar commented on common/collision.h:
```diff
@@ -0,0 +1,195 @@
+
+    static constexpr meter_t inf()
+    {
+        return meter_t{ std::numeric_limits<double>::infinity() };
```
This doesn't compile for me. I've used the ``UNIT_LIB_DEFAULT_TYPE`` macro instead, which lets you get (and set) the unit lib type.

neworderofjamie replied:
maybe you need an extra ``typedef `` or some shit...

alexdewar commented:
Ta for looking through! Would you mind installing SFML so that we can pass the unit tests?
